File: backend/app/profiles/telecom.py
from app.profiles.base import (
    ActionResult,
    ActionSpec,
    OutboundProfile,
    Profile,
    ToolSpec,
    make_greeting,
    object_schema,
    rupees,
)

_PLANS = {
    "Postpaid 599": {"price": 599, "data": "60 GB", "extras": "unlimited calls"},
    "Postpaid 799": {"price": 799, "data": "120 GB", "extras": "unlimited calls and a streaming bundle"},
    "Postpaid 999": {
        "price": 999,
        "data": "250 GB",
        "extras": "unlimited calls, a streaming bundle and international roaming",
    },
}


def _get_usage(data, args):
    usage = data["usage"]

    return {
        "used_gb": usage["used_gb"],
        "limit_gb": usage["limit_gb"],
        "remaining_gb": usage["limit_gb"] - usage["used_gb"],
        "resets_on": usage["resets_on"],
    }


def _get_bill(data, args):
    bill = data["bill"]

    return {"amount": rupees(bill["amount"]), "due_date": bill["due_date"], "status": bill["status"]}


def _get_plan(data, args):
    plan = data["plan"]

    return {
        "plan": plan,
        "monthly_price": rupees(data["plans"][plan]["price"]),
        "renews_on": data["usage"]["resets_on"],
    }


def _list_plans(data, args):
    return {
        "current_plan": data["plan"],
        "plans": [
            {"name": name, "monthly_price": rupees(plan["price"]), "data": plan["data"], "includes": plan["extras"]}
            for name, plan in data["plans"].items()
        ],
    }


def _validate_upgrade(data, args):
    plan = args.get("plan")

    if plan not in data["plans"]:
        return f"I can upgrade you to one of: {', '.join(data['plans'])}."
    if plan == data["plan"]:
        return f"You're already on {plan}."
    return None


def _describe_upgrade(args, data):
    plan = data["plans"][args["plan"]]

    return (
        f"upgrade you from {data['plan']} to {args['plan']}, which is "
        f"{rupees(plan['price'])} a month with {plan['data']}. The new price applies "
        "from your next bill"
    )


def _execute_upgrade(args, data):
    previous = data["plan"]
    data["plan"] = args["plan"]

    return ActionResult(
        result={"previous_plan": previous, "new_plan": args["plan"]},
        summary=f"Plan upgraded from {previous} to {args['plan']}",
        reply=f"Done. You're now on {args['plan']}. The new price applies from your next bill.",
        ref=f"Number {data['number']}",
    )


profile = Profile(
    id="telecom",
    name="Telecom Support",
    counterparty="subscriber",
    persona=(
        "You are the virtual assistant for Vertex Mobile, speaking with the subscriber "
        "Meera Iyer about her data usage, bill, plan renewal and upgrades."
    ),
    greeting=make_greeting(
        "I'm the virtual assistant for Vertex Mobile, speaking with Meera Iyer.",
        "I can help with your data usage, your bill, plan renewal, or an upgrade. "
        "What can I do for you?",
    ),
    next_steps=[
        "Pay the outstanding bill before the due date",
        "Check the new plan benefits in the app",
        "Review usage again before the next renewal",
    ],
    data={
        "subscriber": "Meera Iyer",
        "number": "ending 0210",
        "plan": "Postpaid 599",
        "usage": {"used_gb": 42, "limit_gb": 60, "resets_on": "28 September 2026"},
        "bill": {"amount": 599, "due_date": "25 September 2026", "status": "unpaid"},
        "plans": _PLANS,
    },
    tools={
        "get_usage": ToolSpec(
            name="get_usage",
            topic="Data usage",
            description="Get data used, remaining and when the allowance resets.",
            parameters=object_schema({}),
            run=_get_usage,
            ref=lambda data, args: f"Number {data['number']}",
        ),
        "get_bill": ToolSpec(
            name="get_bill",
            topic="Current bill",
            description="Get the current bill amount, due date and payment status.",
            parameters=object_schema({}),
            run=_get_bill,
            ref=lambda data, args: f"Number {data['number']}",
        ),
        "get_plan": ToolSpec(
            name="get_plan",
            topic="Plan renewal",
            description="Get the current plan, its monthly price and renewal date.",
            parameters=object_schema({}),
            run=_get_plan,
            ref=lambda data, args: f"Number {data['number']}",
        ),
        "list_plans": ToolSpec(
            name="list_plans",
            topic="Available plans",
            description="List the plans available to switch to, with price, data and inclusions.",
            parameters=object_schema({}),
            run=_list_plans,
            ref=lambda data, args: "Plan catalogue",
        ),
    },
    actions={
        "upgrade_plan": ActionSpec(
            name="upgrade_plan",
            topic="Plan upgrade",
            label="Upgrade plan",
            description=(
                "Change the subscriber to a different plan. Use when they ask to upgrade "
                "or switch plans. If they don't say which plan, propose Postpaid 799."
            ),
            parameters=object_schema(
                {"plan": {"type": "string", "enum": list(_PLANS), "description": "Target plan."}},
                required=["plan"],
            ),
            validate=_validate_upgrade,
            describe=_describe_upgrade,
            execute=_execute_upgrade,
        ),
    },
    outbound=OutboundProfile(
        organisation="Vertex Mobile",
        persona=(
            "You are calling a subscriber on behalf of Vertex Mobile about their plan renewal "
            "and data usage."
        ),
        objective=(
            "Tell them their renewal date and how much data they have used, and offer an upgrade "
            "only if they are close to their limit or ask about one."
        ),
        actions=["upgrade_plan"],
    ),
    demo_name="Meera Iyer",
)

File: frontend/src/profiles/telecom.js
import { listSentence, makeGreeting, rupees } from "./shared.js";

const PLANS = {
  "Postpaid 599": { price: 599, data: "60 GB", extras: "unlimited calls" },
  "Postpaid 799": { price: 799, data: "120 GB", extras: "unlimited calls and a streaming bundle" },
  "Postpaid 999": { price: 999, data: "250 GB", extras: "unlimited calls, a streaming bundle and international roaming" },
};

function requestedPlan(text) {
  const match = text.match(/\b(799|999)\b/);

  return match ? `Postpaid ${match[1]}` : null;
}

export default {
  id: "telecom",
  name: "Telecom Support",
  vertical: "Telecom",
  workspaceLabel: "Telecom Support Workspace",
  counterparty: "subscriber",

  headline: "How can I help with your plan?",
  description:
    "Speak naturally with the AI voice agent about your usage, bill, plan renewal or an upgrade.",

  greeting: makeGreeting(
    "I'm the virtual assistant for Vertex Mobile, speaking with Meera Iyer.",
    "I can help with your data usage, your bill, plan renewal, or an upgrade. What can I do for you?"
  ),

  prompts: [
    "How much data have I used?",
    "What's my current bill?",
    "When does my plan renew?",
    "What plans are available?",
    "Upgrade my plan",
  ],

  capabilities: [
    "check your data usage",
    "explain your bill",
    "tell you when your plan renews",
    "upgrade your plan",
  ],

  nextSteps: [
    "Pay the outstanding bill before the due date",
    "Check the new plan benefits in the app",
    "Review usage again before the next renewal",
  ],

  data: {
    subscriber: "Meera Iyer",
    number: "ending 0210",
    plan: "Postpaid 599",
    usage: { usedGb: 42, limitGb: 60, resetsOn: "28 September 2026" },
    bill: { amount: 599, dueDate: "25 September 2026", status: "unpaid" },
    plans: PLANS,
  },

  intents: [
    {
      id: "upgrade",
      topic: "Plan upgrade",
      match: /\b(upgrade|switch|change (?:my )?plan|move to)\b/i,
      propose: (data, text) => {
        const plan = requestedPlan(text) ?? "Postpaid 799";

        if (plan === data.plan) {
          return { reply: `You're already on ${plan}.` };
        }

        return { tool: "upgrade_plan", args: { plan } };
      },
    },
    {
      id: "plans",
      topic: "Available plans",
      tool: "list_plans",
      match: /\b(plans|options|available|offers)\b/i,
      run: (data) => {
        const parts = Object.entries(data.plans).map(
          ([name, plan]) => `${name} at ${rupees(plan.price)} with ${plan.data} and ${plan.extras}`
        );

        return {
          args: {},
          result: data.plans,
          reply: `The plans available are ${listSentence(parts)}. You're on ${data.plan}.`,
          ref: "Plan catalogue",
        };
      },
    },
    {
      id: "usage",
      topic: "Data usage",
      tool: "get_usage",
      match: /\b(data|usage|used|left|remaining|gb)\b/i,
      run: (data) => {
        const { usedGb, limitGb, resetsOn } = data.usage;

        return {
          args: {},
          result: data.usage,
          reply: `You've used ${usedGb} of your ${limitGb} GB, so ${limitGb - usedGb} GB is left. It resets on ${resetsOn}.`,
          ref: `Number ${data.number}`,
        };
      },
    },
    {
      id: "bill",
      topic: "Current bill",
      tool: "get_bill",
      match: /\b(bill|payment|pay|amount|charges)\b/i,
      run: (data) => ({
        args: {},
        result: data.bill,
        reply: `Your current bill is ${rupees(data.bill.amount)}, due on ${data.bill.dueDate}. It is ${data.bill.status}.`,
        ref: `Number ${data.number}`,
      }),
    },
    {
      id: "renewal",
      topic: "Plan renewal",
      tool: "get_plan",
      match: /\b(renew|renewal|validity|expire|expiry|plan)\b/i,
      run: (data) => ({
        args: {},
        result: { plan: data.plan, renews: data.usage.resetsOn },
        reply: `You're on ${data.plan} at ${rupees(data.plans[data.plan].price)} a month. It renews on ${data.usage.resetsOn}.`,
        ref: `Number ${data.number}`,
      }),
    },
  ],

  actions: {
    upgrade_plan: {
      label: "Upgrade plan",
      describe: (args, data) => {
        const plan = data.plans[args.plan];

        return `upgrade you from ${data.plan} to ${args.plan}, which is ${rupees(plan.price)} a month with ${plan.data}. The new price applies from your next bill`;
      },
      execute: (args, data) => {
        const previous = data.plan;

        data.plan = args.plan;

        return {
          result: { previous_plan: previous, new_plan: args.plan },
          summary: `Plan upgraded from ${previous} to ${args.plan}`,
          reply: `Done. You're now on ${args.plan}. The new price applies from your next bill.`,
          ref: `Number ${data.number}`,
        };
      },
    },
  },
};
